src/common/CMakeLists.txt

```diff
@@ -1,191 +1,230 @@
+if(HAVE_MONOTONIC_CLOCK)
+	message(STATUS "Check for monotonic clock - yes")
+	set(MONOTONIC_CLOCK_LIBRARIES ${RT_LIBRARY})
+	set(MONOTONIC_CLOCK_DEFINITIONS "-DHAVE_MONOTONIC_CLOCK")
+else()
+	message(STATUS "Check for monotonic clock - no")
+endif()
 
-	if( MSVC )
-		add_definitions(-D_WINSOCK_DEPRECATED_NO_WARNINGS)
-	endif()
+# Check ENABLE_MEMORY option
+if(ENABLE_MEMORY STREQUAL "system")
+	# use the normal system functions
+else()
+	message(STATUS "Feature not implemented, defaulting to system")
+endif()
 
-	set( COMMON_ADDITIONALL_CPP	"${COMMON_SOURCE_DIR}/winapi.cpp" )
 
-	set( COMMON_ADDITIONALL_HPP	"${COMMON_SOURCE_DIR}/winapi.hpp" )
-endif()
+add_library(minicore EXCLUDE_FROM_ALL)
+add_library(common)
+# add_library(core-tools)
+add_library(common-win INTERFACE)
 
-set( COMMON_ALL_HEADERS
-	"${CMAKE_CURRENT_BINARY_DIR}/version.hpp"
-	"${COMMON_SOURCE_DIR}/cbasetypes.hpp"
-	"${COMMON_SOURCE_DIR}/mmo.hpp"
+if(WIN32)
+	target_link_libraries(common-win INTERFACE "ws2_32.lib")
+
+	if(MSVC)
+		target_compile_definitions(common-win INTERFACE
+			_WINSOCK_DEPRECATED_NO_WARNINGS _CRT_SECURE_NO_DEPRECATE _CRT_NONSTDC_NO_DEPRECATE)
+	endif()
+
+	target_sources(common-win INTERFACE
+		"${CMAKE_CURRENT_SOURCE_DIR}/winapi.cpp"
+		"${CMAKE_CURRENT_SOURCE_DIR}/winapi.hpp"
 	)
 
-add_library(minicore)
+	target_link_libraries(minicore PUBLIC common-win)
+	target_link_libraries(common PUBLIC common-win)
+endif()
 
 target_sources(minicore PRIVATE
-	"${COMMON_SOURCE_DIR}/core.cpp"
-	"${COMMON_SOURCE_DIR}/malloc.cpp"
-	"${COMMON_SOURCE_DIR}/showmsg.cpp"
-	"${COMMON_SOURCE_DIR}/strlib.cpp"
-	${LIBCONFIG_SOURCES} # needed by showmsg.cpp
-	${COMMON_ADDITIONALL_CPP} # needed by Windows
+	"core.cpp"
+	"database.cpp"
+	"des.cpp"
+	"grfio.cpp"
+	"malloc.cpp"
+	"nullpo.cpp"
+	"showmsg.cpp"
+	"strlib.cpp"
+	"utils.cpp"
+)
+
+if(WIN32)
+	target_sources(minicore PRIVATE
+		"cbasetypes.hpp"
+		"core.hpp"
+		"database.hpp"
+		"des.hpp"
+		"grfio.hpp"
+		"malloc.hpp"
+		"nullpo.hpp"
+		"showmsg.hpp"
+		"strlib.hpp"
+		"utils.hpp"
+	)
+	set_target_properties(minicore PROPERTIES FOLDER "Core")
+endif()
+
+target_compile_definitions(minicore PRIVATE
+	"-DMINICORE"
+	${MONOTONIC_CLOCK_DEFINITIONS}
 )
 
 target_include_directories(minicore PUBLIC
-	${COMMON_SOURCE_DIR}
-	${LIBCONFIG_INCLUDE_DIRS}
-	${COMMON_ADDITIONALL_HPP} # needed by Windows
+	"${RA_INCLUDE_DIRS}"
+	"${ZLIB_INCLUDE_DIRS}"
 )
 
+target_link_libraries(minicore PUBLIC
+	${MONOTONIC_CLOCK_LIBRARIES}
+	libconfig
+	ryml
+	${ZLIB_LIBRARIES}
+	yaml-cpp
+)
 
-target_compile_definitions(minicore PRIVATE "-DMINICORE" ${LIBCONFIG_DEFINITIONS})
+target_sources(common PRIVATE
+	"cli.cpp"
+	"conf.cpp"
+	"core.cpp"
+	"database.cpp"
+	"db.cpp"
+	"des.cpp"
+	"ers.cpp"
+	"grfio.cpp"
+	"malloc.cpp"
+	"mapindex.cpp"
+	"md5calc.cpp"
+	"msg_conf.cpp"
+	"nullpo.cpp"
+	"random.cpp"
+	"showmsg.cpp"
+	"socket.cpp"
+	"sql.cpp"
+	"strlib.cpp"
+	"timer.cpp"
+	"utils.cpp"
+	"utilities.cpp"
+)
```

Comment:
```suggestion
)

if(ENABLE_EXTRA_BUILDBOT_CODE)
	file(GLOB COMMON_SOURCES ${CMAKE_CURRENT_SOURCE_DIR}/*.cpp)

	get_target_property(compile_defs common SOURCES)
	list(LENGTH compile_defs cdeflen)
	list(TRANSFORM compile_defs PREPEND ${CMAKE_CURRENT_SOURCE_DIR}/)

	foreach (my_entry IN LISTS COMMON_SOURCES)
		if (NOT my_entry IN_LIST compile_defs)
			message(FATAL_ERROR "${my_entry} is missing in target_sources for common")
		endif()
	endforeach()
endif()
```
Can we add such checks for the (most important) targets? For sources + headers.
Would require a small redesign of your common-win library though.

Reply:
Why would this be needed? If someone forgets to add the file in cmakelists.txt, then it'll fail to build.

Reply:
True, chance is really low that a header or source file will not be included in an already existing file.
May have been my way of thinking of thinking of cmake as the unmaintained compilation tool.
Since it will become our main compilation tool over time, the chances of such files will get lower and lower anyway.